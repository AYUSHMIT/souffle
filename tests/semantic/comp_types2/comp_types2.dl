// Souffle - A Datalog Compiler
// Copyright (c) 2021, The Souffle Developers. All rights reserved
// Licensed under the Universal Permissive License v 1.0 as shown at:
// - https://opensource.org/licenses/UPL
// - <souffle root>/licenses/SOUFFLE-UPL.txt

// Tests parameter instantiations of components
// for user-defined types.

//////////
// Test 1
//////////

.comp B<P> {
  .type T = branch {x:P}
  .decl R(x:T)
}
.init myB = B<number>
myB.R($branch(1)).
.printsize myB.R

//////////
// Test 2
//////////

.comp C<P> {
  .type P <: number
  .decl R(x:P)
}
.init myC = C<myType>
myC.R(1).
.printsize myC.R

//////////
// Test 3
//////////

.comp D<P> {
  .type X <: P
  .decl R(x:X)
}
.init myD = D<number>
myD.R(1).
.printsize myD.R

//////////
// Test 4
//////////

.type myT2 <: number
.comp E<P> {
  .type myT <: number
  .type X = P | myT
  .decl R(x:X)
}
.init myE = E<myT2>
myE.R(1).
.printsize myE.R

