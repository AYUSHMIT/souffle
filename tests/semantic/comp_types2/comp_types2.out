myB.R	1
myC.R	1
myD.R	1
myE.R	1
